File: src/main/java/bad/robot/radiate/teamcity/BasicAuthConfiguration.java
package bad.robot.radiate.teamcity;

import bad.robot.http.CommonHttpClient;
import bad.robot.http.configuration.BasicAuthCredentials;

@Deprecated
class BasicAuthConfiguration {

    private final Server server;
    private final Username username;
    private final Password password;

    public static BasicAuthConfiguration basicAuthConfiguration(TeamCityConfiguration configuration) {
        Username username = configuration.username();
        Password password = configuration.password();
        if (username.equals(new NoUsername()) || password.equals(new NoPassword()))
            return new GuestAuthentication();

        String host = configuration.host();
        Integer port = configuration.port();
        Server server = new Server(host, port);
        return new BasicAuthConfiguration(server, username, password);
    }

    protected BasicAuthConfiguration(Server server, Username username, Password password) {
        this.server = server;
        this.username = username;
        this.password = password;
    }

    public void applyTo(CommonHttpClient client) {
        client.with(BasicAuthCredentials.basicAuth(username.asSimpleHttp(), password.asSimpleHttp(), server.urlFor(() -> "/")));
    }

}

File: src/main/scala/bad/robot/radiate/teamcity/BasicAuthConfiguration.scala
package bad.robot.radiate.teamcity

import bad.robot.http.CommonHttpClient
import bad.robot.http.configuration.BasicAuthCredentials
import bad.robot.radiate.FunctionInterfaceOps.toHypermedia

object BasicAuthConfigurationS {
  def apply(configuration: TeamCityConfigurationS): BasicAuthConfigurationS = {
//    (configuration.username, configuration.password) match {
//      case (NoUsernameS, _) => new GuestAuthenticationS
//      case (_, noPassword) => new GuestAuthenticationS
//    }

    val username = configuration.username
    val password = configuration.password
    if ((username == NoUsernameS) || (password == NoPasswordS))
      return GuestAuthenticationS
    val host = configuration.host
    val port = configuration.port
    val server = new ServerS(host, port)
    new BasicAuthConfigurationS(server, username, password)
  }
}

case class BasicAuthConfigurationS (server: ServerS, username: UsernameS, password: PasswordS) {
  def applyTo(client: CommonHttpClient) {
    client.`with`(BasicAuthCredentials.basicAuth(username.asSimpleHttp, password.asSimpleHttp, server.urlFor("/")))
  }
}